### DISCUSSION TITLE
Testing queries using "connect" and mock Prisma Client

### DISCUSSION THREAD

**Discussion by davidlag0** [prisma/prisma#11193] [Q&A] [answered]
Hi!

I followed the documentation here (https://www.prisma.io/docs/guides/testing/unit-testing) to create a mock Prisma Client to create unit tests. It works well if I remove the use of `connect` in my `prisma.model.create()` call and add a line to specify the record ID to link to (`authorId: 1,`). If I leave `connect`, then I get this error message:

```
Invalid `prisma.task.create()` invocation:

      An operation failed because it depends on one or more records that were required but not found. No 'User' record(s) (needed to inline the relation on 'Task' record(s)) was found for a nested connect on one-to-many relation 'TaskToUser'.
```

These are my Prisma models:

```prisma
model User {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now()) @map(name: "created_at")
  updatedAt DateTime @updatedAt @map(name: "updated_at")
  email     String   @unique
  name      String?
  role      Role     @default(USER)
  tasks     Task[]
  @@map(name: "users")
}

model Task {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now()) @map(name: "created_at")
  updatedAt DateTime @updatedAt @map(name: "updated_at")
  name      String   @db.VarChar(255)
  checked   Boolean  @default(false)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
}

enum Role {
  USER
  ADMIN
}
```

This is my `prisma.task.create()` call:
```js
      const result = await prisma.task.create({
        data: {
          name: name,
          author: { connect: { email: session?.user?.email } },
        },
      });
```

This is the test I'm trying to get working with the use of `connect`:
```js
  test("returns task details when logged in and POST'ing a new task name", async () => {
    const { req, res } = createMocks({
      method: "POST",
      body: { name: "Test Task" },
    });

    getSession.mockReturnValue({
      user: {
        name: "Test Author",
        email: "testauthor@test.com",
        image: "testimage",
      },
      expires: "1",
    });

    const task = {
      name: "Test Task",
      checked: false,
      authorId: 1,
    };

    prismaMock.task.create.mockResolvedValue(task);

    await handle(req, res);

    expect(res._getStatusCode()).toBe(200);
    expect(JSON.parse(res._getData())).toEqual(expect.objectContaining(task));
  });
```

Any pointer regarding what else needs to be mocked to be able to use `connect` would be greatly appreciated. I have tried to do things like call `prisma.user.create()` with data hoping that would feed the Prisma Client with the data that can then be reused later but it isn't working. I also came across GitHub issues with asks for an in-memory database with SQLite and so on. Just knowing what to mock to get the test passing would be great.

Thank you very much!

**Comment by kaushalrajbacancy** [comment]
Could you please share more information? The entire code is the same. But, if I add the following code, it gets crashed and If I remove the code it reaches the res.json() in the API handler.

`const result = await prisma.task.create({
        data: {
          name: name,
         },
      });`

**Reply by pantharshit00** [reply]
Hello @davidlag0,

Sorry for late followup here. Did you get a chance to review the above suggestion? Let us know if this issue is still relevant to you.

**Reply by davidlag0** [reply]
I'm taking another look at this and I need more information about your suggestion @kaushalrajbacancy . If I remove the line of code you suggest removing, then the task is not created, which goes against the idea of that code block: a POST request is made to the endpoint to create a new task.

**Reply by davidlag0** [reply]
@pantharshit00 @kaushalrajbacancy In the sake of avoiding losing time on this, I believe I found the problem in relation to the error message I'm getting and it seems to be a misunderstanding on my end. I believe I'm just missing an initial entry in my database (a test user entry in my Users table). I appreciate you responded and brought up an idea, which ended up driving me to investigate more and learn. I'm currently working to resolve this with an initial DB seed and I'll report back to confirm it works when I'm done.

**Reply by davidlag0** [reply]
@pantharshit00 @kaushalrajbacancy I just finished making my changes and I confirm I misunderstood I just needed to seed my database for add the missing record that was causing the error message above about the User record that cannot be found. Thanks again for your help!

**Comment by davidlag0** [comment]
@pantharshit00 Sorry, I didn't get a chance to try to above suggestion. I ended up changing my code to make it work differently:
https://github.com/davidlag0/todo-nextjs/blob/main/__tests__/unit/api/tasks/index.test.js#L68-L85

I can try to revisit it and see if the problem I had with connect is now resolved.

Thank you!

**Comment by davidlag0** [comment, marked as answer]
The solution was just to seed the database using the information here: https://www.prisma.io/docs/guides/database/seed-database

With this in place, I update my npm scripts to run the seed command after the database is initialized and the call using `connect` worked as the missing record was now there!

Enabling debug logs (`export DEBUG="prisma*"`) following the instructions here (https://www.prisma.io/docs/concepts/components/prisma-client/debugging#setting-the-debug-environment-variable) helped me get more information to troubleshoot the issue.

Thanks again for your responses to help provide me the drive to find the problem!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
